scripts/plugins/extauth-hook-AD.py

```diff
@@ -135,77 +138,88 @@ class StaticPam(ADConfig):
             else:
                 ad_pam_module = "pam_winbind.so"
-            content = self.ad_pam_format.format(ad_module=ad_pam_module)
+            content = self.ad_pam_format.format(
+                ad_module=ad_pam_module, list_sep=self.list_sep,
+                field_sep=self.field_sep, conf_file=self.conf_file)
         else:
             content = self.no_ad_pam
         self._lines = content.split("\n")
 
 
 class DynamicPam(ADConfig):
     #pylint: disable=too-few-public-methods
-    """Class manage /etc/pam.d/hcp_users which permit pool admin ssh"""
+    """Class manage /etc/security/hcp_access.conf which permit pool admin ssh"""
+
+    # the format for pam_access is permission:users:origins
+    # details, refer to /etc/security/access.conf
+    CONF_TEMP = "{perm}{sep}{name}{sep}{origins}"
 
-    def __init__(self, session, arg, ad_enabled=True):
-        super(DynamicPam, self).__init__("/etc/pam.d/hcp_users", session, arg, ad_enabled,
+    def __init__(self, session, args, ad_enabled=True):
+        super(DynamicPam, self).__init__("/etc/security/hcp_access.conf", session, args, ad_enabled,
                                          load_existing=False)
+        self.field_sep = args.get("field_sep", "`")
 
     def _apply_to_cache(self):
         # AD is not enabled, just return as the configure file will be removed during install
         if not self._ad_enabled:
             return
         # Rewrite the PAM SSH config using the latest info from Active Directory
         # and the list of subjects from xapi
-        subjects = self._session.xenapi.subject.get_all()
-        admin_role = self._session.xenapi.role.get_by_name_label(
-            'pool-admin')[0]
-        # Add each subject which contains the admin role
-        for opaque_ref in subjects:
-            subject_rec = self._session.xenapi.subject.get_record(opaque_ref)
-            if admin_role in subject_rec['roles']:
-                self._add_subject(subject_rec)
-
-    def _format_item(self, item):
+        try:
+            subjects = self._session.xenapi.subject.get_all()
+            admin_role = self._session.xenapi.role.get_by_name_label(
+                'pool-admin')[0]
+            # Add each subject which contains the admin role
+            for opaque_ref in subjects:
+                subject_rec = self._session.xenapi.subject.get_record(
+                    opaque_ref)
+                if admin_role in subject_rec['roles']:
+                    self._add_subject(subject_rec)
+        except Exception:  # pylint: disable=broad-except
+            logger.exception("Failed permit subjects")
+        finally:
+            # deny all others by default
+            conf = self.CONF_TEMP.format(
+                perm="-", sep=self.field_sep, name="ALL", origins="ALL")
+            self._lines.append(conf)
+            self._lines.append("")  # Require ending line sep
+
+    def _format_item(self, item, is_group):
         space_replacement = "+"
```

Comment:
It would be useful to check that the user/group name doesn't contain any of the field separators. Even if currently AD would forbid those characters:
* the separator is configurable in xapi.conf, so someone could accidentally choose an unsafe one
* future versions of AD may change what characters they allow, and if someone manages to create an AD user with one of the escape chars in their user/group name they'd be able to elevate privileges.

If the user/group contains the field separator then see whether it is possible to escape it, if not then we should reject adding that user/group to XAPI and for an already existing user/group we should log+skip the entry here.

Reply:
@edwintorok  we forbid add separators to avoid injection attack, 
However, we will not forbid add subject with separators, given
- if Microsoft allow it, then we allow it.
- we forbid add to pam configuration, which disable ssh, However, it should not affect other functions like  add users, user login with SDK, CHC.   
We log warning to remind user update separators if they insist to.